src/osmdata.cpp

```diff
@@ -389,16 +393,41 @@ progress_display_t osmdata_t::process_file(osmium::io::File const &file,
     return handler.progress();
 }
 
-void osmdata_t::process_stage1b() const
+void osmdata_t::process_stage1bc() const
 {
-    if (m_dependency_manager->has_pending()) {
-        multithreaded_processor proc{m_conninfo, m_mid, m_outs,
-                                     (std::size_t)m_num_procs};
+    if (!m_dependency_manager->has_pending()) {
+        return;
+    }
+
+    multithreaded_processor proc{m_conninfo, m_mid, m_outs,
+                                 (std::size_t)m_num_procs};
+
+    // stage 1b processing
+    proc.process_ways(m_dependency_manager->get_pending_way_ids());
+    proc.process_relations(m_dependency_manager->get_pending_relation_ids());
+    proc.merge_expire_trees();
+
+    // If none of the outputs has a check_relation defined, we can skip the
+    // stage 1c processing.
+    if (std::none_of(m_outs.cbegin(), m_outs.cend(),
+                     [](std::shared_ptr<output_t> const &out) {
+                         return out->has_stage2_processing();
+                     })) {
+        return;
+    }
 
-        proc.process_ways(m_dependency_manager->get_pending_way_ids());
+    // stage 1c processing
+    for (auto &out : m_outs) {
+        if (out->has_stage1c_pending()) {
+            for (auto const id : out->get_stage1c_way_ids()) {
+                m_dependency_manager->way_changed(id);
+            }
+        }
+    }
+
+    if (m_dependency_manager->has_pending()) {
```

Comment:
We not only should disable `add_rows()`, we shouldn't even call `process_relation()`, but only do the member way marking. I forgot about that here, but it will be in the next version of the PR.